File: leetcode/go/987.vertical-order-traversal-of-a-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func verticalTraversal(root *TreeNode) [][]int {
    vot := map[int][][]int{}
    dfs(root, 0, 0, vot)
    keys := make([]int, 0, len(vot))
    ans := make([][]int, 0, len(vot))
    for k := range vot {
        keys = append(keys, k)
    }
    slices.Sort(keys)
    for _, k := range keys {
        v := vot[k]
        sort.Slice(v, func(i, j int) bool {
            x, y := v[i], v[j]
            if x[0] == y[0] {
                return x[1] < y[1]
            }
            return x[0] < y[0]
        })
        vals := make([]int, 0, len(v))
        for _, node_val := range v {
            vals = append(vals, node_val[1])
        }
        ans = append(ans, vals)
    }
    return ans
}

func dfs(root *TreeNode, x int, y int, vot map[int][][]int) {
    if root == nil {
        return
    }
    if _, ok := vot[y]; ok {
        vot[y] = append(vot[y], []int{x, root.Val})
    } else {
        v := [][]int{}
        v = append(v, []int{x, root.Val})
        vot[y] = v
    }
    dfs(root.Left, x+1, y-1, vot)
    dfs(root.Right, x+1, y+1, vot)
}

File: leetcode/rust-leetcode/src/987.vertical-order-traversal-of-a-binary-tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::BTreeMap;
type TypeNode = Option<Rc<RefCell<TreeNode>>>;
impl Solution {
    pub fn vertical_traversal(root: TypeNode) -> Vec<Vec<i32>> {
        fn dfs(root: &TypeNode, pose: (i32, i32), vot: &mut BTreeMap<i32, Vec<(i32, i32)>>) {
            match root {
                None => {},
                Some(node) => {
                    let node = node.borrow();
                    vot.entry(pose.1).or_insert(Vec::new()).push((pose.0, node.val));
                    dfs(&node.left, (pose.0+1, pose.1-1), vot);
                    dfs(&node.right, (pose.0+1, pose.1+1), vot);
                }
            }
        }
        let mut vot = BTreeMap::new();
        dfs(&root, (0, 0), &mut vot);
        let mut ans = Vec::with_capacity(vot.len());
        for (k, mut v) in vot.into_iter() {
            v.sort();
            let vertical = v.iter().map(|&(x, node_val)| node_val).collect::<Vec<_>>();
            ans.push(vertical.to_vec());
        }
        ans
    }
}

